lib/renderer/window-setup.ts

```diff
@@ -0,0 +1,281 @@
+  }
+
+  constructor (private ipcRenderer: Electron.IpcRenderer, private guestId: any) {
+    this._location = new LocationProxy(ipcRenderer, guestId)
+
+    ipcRenderer.once(`ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_CLOSED_${guestId}`, () => {
+      removeProxy(guestId)
+      this.closed = true
+    })
+  }
+
+  public close () {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_CLOSE', this.guestId)
+  }
+
+  public focus () {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_METHOD', this.guestId, 'focus')
+  }
+
+  public blur () {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_METHOD', this.guestId, 'blur')
+  }
+
+  public print () {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WEB_CONTENTS_METHOD', this.guestId, 'print')
+  }
+
+  public postMessage (message: any, targetOrigin: any) {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_POSTMESSAGE', this.guestId, message, toString(targetOrigin), window.location.origin)
+  }
+
+  public eval (...args: any[]) {
+    this.ipcRenderer.send('ELECTRON_GUEST_WINDOW_MANAGER_WEB_CONTENTS_METHOD', this.guestId, 'executeJavaScript', ...args)
+  }
+}
+
+const windowSetup = (
+  ipcRenderer: Electron.IpcRenderer, guestInstanceId: number, openerId: number, isHiddenPage: boolean, usesNativeWindowOpen: boolean
+) => {
+  if (guestInstanceId == null) {
+    // Override default window.close.
+    window.close = function () {
+      ipcRenderer.sendSync('ELECTRON_BROWSER_WINDOW_CLOSE')
+    }
+  }
+
+  if (!usesNativeWindowOpen) {
+    // Make the browser window or guest view emit "new-window" event.
+    (window as any).open = function (url?: string, frameName?: string, features?: string) {
+      if (url != null && url !== '') {
+        url = resolveURL(url)
+      }
+      const guestId = ipcRenderer.sendSync('ELECTRON_GUEST_WINDOW_MANAGER_WINDOW_OPEN', url, toString(frameName), toString(features))
+      if (guestId != null) {
+        return getOrCreateProxy(ipcRenderer, guestId)
+      } else {
+        return null
+      }
+    }
+
+    if (openerId != null) {
+      window.opener = getOrCreateProxy(ipcRenderer, openerId)
+    }
+  }
+
+  // But we do not support prompt().
+  window.prompt = function () {
+    throw new Error('prompt() is and will not be supported.')
+  }
+
+  ipcRenderer.on('ELECTRON_GUEST_WINDOW_POSTMESSAGE', function (
+    _event: Event, sourceId: number, message: any, sourceOrigin: string
+  ) {
+    // Manually dispatch event instead of using postMessage because we also need to
+    // set event.source.
+    const messageEvent = new MessageEvent('message', {
+      data: message,
+      origin: sourceOrigin,
+      source: getOrCreateProxy(ipcRenderer, sourceId) as any as Window,
+      bubbles: false,
+      cancelable: false
+    })
+
+    window.dispatchEvent(messageEvent as MessageEvent)
```

Comment:
This `as` cast seems strange, it's `new MessageEvent`, shouldn't it know that it is indeed a `MessageEvent`